go/pkg/filemetadata/cache.go

```diff
@@ -17,11 +18,17 @@ type fmCache struct {
 	Backend     *cache.Cache
 	cacheHits   uint64
 	cacheMisses uint64
+
+	// Semaphoe to limit the number of parallel hash calculation.
+	computeSema chan struct{}
 }
 
 // NewSingleFlightCache returns a singleton-backed in-memory cache, with no validation.
 func NewSingleFlightCache() Cache {
-	return &fmCache{Backend: cache.GetInstance()}
+	return &fmCache{
+		Backend:     cache.GetInstance(),
+		computeSema: make(chan struct{}, runtime.NumCPU()),
```

Comment:
Did you experiment with going above num CPUs? I imagine you could handle more parallelism than num CPUs.

Reply:
No, but I think this is CPU intensive and we won't get more benefit with higher concurrency?

Reply:
There is an IO component that may allow higher parallelism I would imagine. There's a syscall to stat the file, and then reading to digest.

Reply:
I think having parallelism be an argument for `NewSingleFlightCache` can't hurt. We can come up with a heuristic for the optimal value depending on OS and CPU separately.

Reply:
In our issue, I'm suspecting cpu usage from hash calculation for large files is dominant part.
So I choose cpu num as default not to cause many context switch there.

And I added option to specify parallelism in case user needs to change.